### DISCUSSION TITLE
What is the typical way to persist currency?

### DISCUSSION THREAD

**Discussion by johnhooks** [dinerojs/dinero.js#494] [Q&A] [answered]
> Dinero.js tracks those amendments and updates currencies accordingly. Such changes are, by definition, breaking. To preserve fixed versioning, the library's entry point doesn't export updated currencies. In other words, if you import currencies from the path @dinero.js/currencies, they won't change.

[Using Built In Currencies](https://v2.dinerojs.com/docs/core-concepts/currency#using-built-in-currencies)

I understand that I can rely on the currencies not to change, unless upgrading to a breaking change version of dinero. But what is the typical way to persist the currency of a dinero object, like in a database, so it can be retrieved later and also insured that the appropriate currency exponent and base always be used?

Should the whole snapshot of the dinero object be persisted, with its currency, to ensure its value is always interpreted correctly?

```sql
CREATE TABLE "ExampleAmount" (
    "amount" INTEGER,
    "code" VARCHAR(3);
    "base" INTEGER,
    "exponent" INTEGER,
    "scale" INTEGER
);
```

I get that all applications will have different needs, but I just want to make sure I am on the right track, and not over thinking it.

**Comment by sarahdayan** [comment, marked as answer]
Hey @johnhooks!

>I understand that I can rely on the currencies not to change, unless upgrading to a breaking change version of dinero.

What the documentation means is that, if a currency changes (for example, if the U.S. dollar suddenly becomes 1000 cents in a dollar), changing it in Dinero.js is a breaking change in nature, which is why there's a different importing strategy to avoid breakages.

>But what is the typical way to persist the currency of a dinero object, like in a database, so it can be retrieved later and also insured that the appropriate currency exponent and base always be used?

There's a [whole guide](https://v2.dinerojs.com/docs/guides/transporting-and-restoring) on this topic. It's good practice to store a full snapshot of Dinero objects in database, which is why the API exposes a [`toSnapshot`](https://v2.dinerojs.com/docs/api/formatting/to-snapshot) function.

**Reply by johnhooks** [reply]
Thank you for the validation that I wasn't misunderstanding.

In the example, [Serializing an object](https://v2.dinerojs.com/docs/guides/transporting-and-restoring#serializing-an-object), does a Dinero object have a `toString` method that uses `toSnapshot`? Or does the fetch call have a bug?

```js 
fetch('http://example.org/api/products', {
  method: 'POST',
  body: JSON.stringify(product),
});
```

Should it be above, as in the example, or the snippet below?

```js 
fetch('http://example.org/api/products', {
  method: 'POST',
  body: JSON.stringify(toSnapshot(product)),
});
```

Its also the same in [Handling arbitrary precision amounts](https://v2.dinerojs.com/docs/guides/transporting-and-restoring#handling-arbitrary-precision-amounts). Both examples import `toSnapshot` but don't use it.

**Reply by johnhooks** [reply]
I read the documentation for [JSON.stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify). I found out it will call the `toJSON` method on a dinero object, and `toSnapshot` also does the same thing.

**Reply by sarahdayan** [reply]
Yep, the import in the docs is unused.
